StackCalculator/Program.cs

```diff
@@ -0,0 +1,238 @@
+﻿/// <summary>
+/// Interface for the classic stack.
+/// </summary>
+/// <typeparam name="T"></typeparam>
+public interface IStack<T>
+{
+    /// <summary>
+    /// Pushes a value onto the stack.
+    /// </summary>
+    /// <param name="value">The value to push onto the stack.</param>
+    public void Push(T value);
+
+    /// <summary>
+    /// Pops the top value off the stack and returns it.
+    /// </summary>
+    /// <returns>The top value of the stack.</returns>
+    public T Pop();
+
+    /// <summary>
+    /// Getter for stack size.
+    /// </summary>
+    /// <returns>Stack size.</returns>
+    public int Size { get; }
+}
+
+/// <summary>
+/// Implementation of a stack using a list.
+/// </summary>
+/// <typeparam name="T"></typeparam>
+public class StackOnList<T> : IStack<T>
+{
+    private int size;
+    private readonly List<T> list;
+
+    /// <summary>
+    /// StackOnList constructor.
+    /// </summary>
+    public StackOnList()
+    {
+        size = 0;
+        list = new List<T>();
+    }
+    public void Push(T value)
+    {
+        list.Add(value);
+        size++;
+    }
+    public T Pop()
+    {
+        if (size > 0)
+        {
+            T returningValue = list[^1];
+            size--;
+            list.RemoveAt(size);
+
+            return returningValue;
+        }
+        else
+        {
+            throw new InvalidOperationException("Stack in empty");
+        }
+    }
+    public int Size
+    {
+        get { return size; }
+    }
+}
+
+/// <summary>
+/// Implementation of a stack using an array.
+/// </summary>
+/// <typeparam name="T"></typeparam>
+public class StackOnArray<T> : IStack<T>
+{
+    private int size;
+    private T[] array;
+
+    /// <summary>
+    /// StackOnArray constructor.
+    /// </summary>
+    /// <param name="capacity">Maximum size of stack.</param>
+    public StackOnArray()
+    {
+        size = 0;
+        array = new T[32];
+    }
+
+    public void Push(T value)
+    {
+        if (size >= array.Length)
+        {
+            T[] tempArray = new T[array.Length * 2];
+            Array.Copy(array, tempArray, array.Length);
+            array = tempArray;
+            array[size] = value;
+        }
+
+        array[size] = value;
+        size++;
+    }
+
+    public T Pop()
+    {
+        if (size > 0)
+        {
+            size--;
+            T returningValue = array[size];
+            array[size] = default;
+            return returningValue;
+        }
+        else
+        {
+            throw new InvalidOperationException("Stack is empty");
+        }
```

Comment:
Можно сделать так:
```suggestion
        if (size == 0)
        {
            throw new InvalidOperationException("Stack is empty");
        }

        size--;
        T returningValue = array[size];
        array[size] = default;
        return returningValue;
```
И тогда содержательная часть функции не будет убегать вправо